pyiron_base/generic/util.py

```diff
@@ -57,3 +61,119 @@ def static_isinstance(obj, obj_type):
         return obj_type in obj_class_lst
     else:
         raise TypeError()
+
+
+class Deprecator:
+    """
+    Decorator class to mark functions and methods as deprecated with a uniform
+    warning message at the time the function is called.  The message has the
+    form
+
+        {function_name} is deprecated: {message}. It is not guaranteed to be in
+        service after {version}.
+
+    unless `pending=True` was given.  Then the message will be
+
+        {function_name} will be deprecated in a future version: {message}.
+
+    If message and version are not initialized or given during the decorating
+    call the respective parts are left out from the message.
+
+    >>> deprecate = Deprecator()
+    >>> @deprecate
+    ... def foo(a, b):
+    ...     pass
+    >>> foo(1, 2)
+    DeprecationWarning: __main__.foo is deprecated
+
+    >>> @deprecate("use bar() instead")
+    ... def foo(a, b):
+    ...     pass
+    >>> foo(1, 2)
+    DeprecationWarning: __main__.foo is deprecated: use bar instead
+
+    >>> @deprecate("use bar() instead", version="0.4.0")
+    ... def foo(a, b):
+    ...     pass
+    >>> foo(1, 2)
+    DeprecationWarning: __main__.foo is deprecated: use bar instead.  It is not
+    guaranteed to be in service in vers. 0.4.0
+
+    >>> deprecate = Deprecator(message="pyiron says no!", version="0.5.0")
+    >>> @deprecate
+    ... def foo(a, b):
+    ...     pass
+    >>> foo(1, 2)
+    DeprecationWarning: __main__.foo is deprecated: pyiron says no!  It is not
+    guaranteed to be in service in vers. 0.5.0
+    """
+
+    def __init__(self, message=None, version=None, pending=False):
+        """
+        Initialize default values for deprecation message and version.
+
+        Args
```

Comment:
Formatting is weird here. Should be more like:

```python
Args:
    message (str): Default deprecation message.
    ...
```

Also, I don't have the link with me, but @samwaseda and I decided to thumb our noses at PEP8 and go for the PyCharm default of 120 characters/line. It looks like some of the breaks here might be to accommodate a shorter line.

Reply:
NB: Actually I agreed that 80 was too short but I'm now sticking to 100 characters because my vim default has 200+ characters and with a vertically split window it nicely accommodates an entire line... This being said, I'm perfectly comfortable with other people using 120 characters as long as they accept my default of 100 characters.

Reply:
I generally prefer shorter lines for vertical splits, but will accommodate anything that we agree on.  I changed my project config to 120 now, if that's what we're going with.

Reply:
But I still want the extra newlines and horizontal bars removed 😉

Reply:
Ah, I got my markup confused, I thought the '----' is necessary for the sphinx argument list, but it's ':'.